File: server/routes/attendance.py
# server/routes/attendance.py

import sqlite3
from fastapi import APIRouter, HTTPException # type: ignore
from fastapi.responses import FileResponse # type: ignore
from pathlib import Path

from server.models.schemas import AttendanceRequest
from server.security import (
    validate_name,
    validate_roll,
    validate_session_code,
    ValidationError,
)
from server.services.session_service import session_service
from server.services.db_service import db_service

router = APIRouter()

# -------------------------
# Serve entry.html
# -------------------------
BASE_DIR = Path(__file__).resolve().parent.parent.parent  # presenz root

@router.get("/")
def serve_entry():
    html_file = BASE_DIR / "client" / "entry.html"
    if not html_file.exists():
        raise HTTPException(status_code=404, detail="Entry form not found")
    return FileResponse(html_file)


# -------------------------
# Submit attendance
# -------------------------
@router.post("/submit")
def submit_attendance(payload: AttendanceRequest):
    try:
        # -------------------------
        # Validate Input
        # -------------------------
        name = validate_name(payload.name)
        roll = validate_roll(payload.roll)

        validate_session_code(
            payload.session_code,
            session_service.get_session_code,
        )

        table_name = session_service.get_table_name

        # -------------------------
        # Check submission limit
        # -------------------------
        if not session_service.can_accept_submission():
            return {
                "status": "closed",
                "message": "Attendance session closed",
            }

        # -------------------------
        # Insert Attendance
        # -------------------------
        db_service.insert_attendance(
            table_name=table_name,
            name=name,
            roll=roll,
        )
        session_service.increment_count()
        # If just reached max, print debug once
        if session_service.is_full():
            print("+------------------------------------------------------------+")
            print(f"| [DEBUG] All {session_service._max_count} responses submitted.                         |")
            print("+------------------------------------------------------------+")

        return {
            "status": "success",
            "message": "Attendance recorded successfully",
        }

    # Duplicate roll (UNIQUE constraint)
    except sqlite3.IntegrityError:
        raise HTTPException(
            status_code=409,
            detail="Roll number already submitted",
        )

    # Validation errors
    except ValidationError as e:
        raise HTTPException(
            status_code=400,
            detail=str(e),
        )

    # Any other unexpected error
    except Exception as e:
        print("[ERROR]", e)
        raise HTTPException(
            status_code=500,
            detail="Internal server error",
        )

File: internal/routes/attendance.go
// internal/routes/attendance.go

package routes

import (
	"embed"
	"encoding/json"
	"net/http"
	"strings"

	"github.com/3rr0r-505/Presenz/internal/config"
	"github.com/3rr0r-505/Presenz/internal/models"
	"github.com/3rr0r-505/Presenz/internal/security"
	"github.com/3rr0r-505/Presenz/internal/services"
)

type AttendanceHandler struct {
	Cfg     *config.Config
	Session *services.Session
	DB      *services.DB
	Entry   embed.FS
}

// ServeEntry handles GET / — serves the embedded entry.html form.
func (h *AttendanceHandler) ServeEntry(w http.ResponseWriter, r *http.Request) {
	data, err := h.Entry.ReadFile("entry.html")
	if err != nil {
		writeError(w, http.StatusNotFound, "Entry form not found")
		return
	}
	w.Header().Set("Content-Type", "text/html; charset=utf-8")
	w.Write(data)
}

// SubmitAttendance handles POST /submit.
func (h *AttendanceHandler) SubmitAttendance(w http.ResponseWriter, r *http.Request) {
	var payload models.AttendanceRequest
	if err := json.NewDecoder(r.Body).Decode(&payload); err != nil {
		writeError(w, http.StatusBadRequest, "Invalid request body")
		return
	}

	name, err := security.ValidateName(payload.Name, h.Cfg.Security.MaxNameLength)
	if err != nil {
		writeError(w, http.StatusBadRequest, err.Error())
		return
	}

	roll, err := security.ValidateRoll(payload.Roll, h.Cfg.Security.MaxRollLength)
	if err != nil {
		writeError(w, http.StatusBadRequest, err.Error())
		return
	}

	if err := security.ValidateSessionCode(payload.SessionCode, h.Session.GetSessionCode()); err != nil {
		writeError(w, http.StatusBadRequest, err.Error())
		return
	}

	tableName := h.Session.GetTableName()

	if !h.Session.TryAcceptSubmission() {
		writeJSON(w, http.StatusOK, models.StatusResponse{
			Status:  "closed",
			Message: "Attendance session closed",
		})
		return
	}

	if err := h.DB.InsertAttendance(tableName, name, roll); err != nil {
		if isUniqueConstraintErr(err) {
			h.Session.DecrementCount()
			writeError(w, http.StatusConflict, "Roll number already submitted")
			return
		}
		h.Session.DecrementCount()
		// unexpected DB failure — don't leak err details to client
		writeError(w, http.StatusInternalServerError, "Internal server error")
		return
	}

	if h.Session.IsFull() {
		println("+------------------------------------------------------------+")
		println("| [DEBUG] All responses submitted.                            |")
		println("+------------------------------------------------------------+")
	}

	writeJSON(w, http.StatusOK, models.StatusResponse{
		Status:  "success",
		Message: "Attendance recorded successfully",
	})
}

// isUniqueConstraintErr checks whether a DB error is a UNIQUE constraint
// violation on roll. modernc.org/sqlite doesn't expose a clean typed
// error for this, so it's a string check against the driver's message.
func isUniqueConstraintErr(err error) bool {
	return strings.Contains(err.Error(), "UNIQUE constraint")
}

func writeJSON(w http.ResponseWriter, status int, body any) {
	w.Header().Set("Content-Type", "application/json")
	w.WriteHeader(status)
	json.NewEncoder(w).Encode(body)
}

func writeError(w http.ResponseWriter, status int, detail string) {
	writeJSON(w, status, models.ErrorResponse{Detail: detail})
}
